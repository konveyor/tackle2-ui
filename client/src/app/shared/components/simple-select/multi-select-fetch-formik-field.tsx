import React from "react";
import { FieldHookConfig, useField } from "formik";

import {
  ISimpleSelectFetchProps,
  SimpleSelectFetch,
} from "./simple-select-fetch";

export interface IMultiSelectFetchFormikFieldProps {
  fieldConfig: FieldHookConfig<any>;
  selectConfig: Omit<ISimpleSelectFetchProps, "value" | "onChange" | "onClear">;
  isEqual: (a: any, b: any) => boolean;
}

export const MultiSelectFetchFormikField: React.FC<
  IMultiSelectFetchFormikFieldProps
> = ({ fieldConfig, selectConfig, isEqual }) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelectFetch
      value={field.value}
      onChange={(selection) => {
        const currentValue = field.value || [];

        let nextValue: any[];
        const elementExists = currentValue.find((f: any) => {
          return isEqual(f, selection);
        });

        if (elementExists) {
          nextValue = currentValue.filter((f: any) => !isEqual(f, selection));
        } else {
          nextValue = [...currentValue, selection];
        }

        helpers.setValue(nextValue);
      }}
      onClear={() => helpers.setValue(undefined)}
      {...selectConfig}
    />
  );
};
