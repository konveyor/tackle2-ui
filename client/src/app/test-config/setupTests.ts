import "@testing-library/jest-dom";
import { server } from "@mocks/server";

const mockInitialized = false;

jest.mock("@react-keycloak/web", () => {
  const originalModule = jest.requireActual("@react-keycloak/web");
  return {
    ...originalModule,
    useKeycloak: () => [mockInitialized],
  };
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: () => ({
    pathname: "localhost:3000/example/path",
  }),
}));

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

// Uncomment the following to see request logging in the console:

// server.events.on("request:start", (req) => {
//   console.log(`Handling a request to ${req.url.href}`);
// });

// server.events.on("request:match", (req) => {
//   console.log(`Request to ${req.url.href} was matched with a handler`);
// });

// server.events.on("request:unhandled", (req) => {
//   console.warn(`Request to ${req.url.href} was not handled`);
// });
