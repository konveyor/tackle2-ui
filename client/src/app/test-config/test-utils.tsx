import React, { FC, ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { Application, Archetype, Assessment } from "@app/api/models";
import { RenderHookOptions, renderHook } from "@testing-library/react-hooks";

import { createContext, useContext } from "react";

const QueryClientContext = createContext<QueryClient | undefined>(undefined);

const AllTheProviders: FC<{
  children: React.ReactNode;
  queryClient?: QueryClient;
}> = ({ children, queryClient }) => {
  const internalQueryClient =
    queryClient ||
    new QueryClient({
      defaultOptions: {
        queries: {
          retry: false,
          cacheTime: 1000,
        },
      },
    });
  return (
    <QueryClientProvider client={internalQueryClient}>
      {children}
    </QueryClientProvider>
  );
};

export const useQueryClientContext = () => {
  const context = useContext(QueryClientContext);
  if (context === undefined) {
    throw new Error(
      "useQueryClientContext must be used within a QueryClientContext.Provider"
    );
  }
  return context;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = <TProps, TResult>(
  callback: (props: TProps) => TResult,
  options?: Omit<RenderHookOptions<TProps>, "wrapper"> & {
    queryClient?: QueryClient;
  }
) => {
  const { queryClient, ...rest } = options || {};
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
    <AllTheProviders queryClient={queryClient}>{children}</AllTheProviders>
  );

  return renderHook(callback, { wrapper: Wrapper as React.FC, ...rest });
};

export * from "@testing-library/react";

export { customRender as render };
export { customRenderHook as renderHook };

export const createMockAssessment = (
  overrides: Partial<Assessment> = {}
): Assessment => {
  return {
    id: Math.random(),
    name: "Default name",
    description: "Default description",
    required: true,
    ...overrides,
  } as Assessment;
};

export const createMockApplication = (overrides: Partial<Application> = {}) => {
  return {
    id: Math.random(),
    name: "Default name",
    description: "Default description",
    ...overrides,
  } as Application;
};

export const createMockArchetype = (overrides: Partial<Archetype> = {}) => {
  return {
    id: Math.random(),
    name: "Default name",
    description: "Default description",
    ...overrides,
  } as Archetype;
};
