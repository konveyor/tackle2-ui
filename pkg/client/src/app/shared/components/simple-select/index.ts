export * from "./simple-select";
export * from "./simple-select-fetch";
export * from "./single-select-fetch-formik-field";
export * from "./multi-select-fetch-formik-field";
export * from "./single-select-formik-field";
export * from "./single-select-fetch-optionvalue-formik-field";
export * from "./multi-select-fetch-optionvalue-formik-field";
export * from "./single-select-optionvalue-formik-field";
