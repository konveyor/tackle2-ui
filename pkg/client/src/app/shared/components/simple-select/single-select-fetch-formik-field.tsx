import React from "react";
import { FieldHookConfig, useField } from "formik";

import {
  ISimpleSelectFetchProps,
  SimpleSelectFetch,
} from "./simple-select-fetch";

export interface ISingleSelectFetchFormikFieldProps {
  fieldConfig: FieldHookConfig<any>;
  selectConfig: Omit<ISimpleSelectFetchProps, "value" | "onChange" | "onClear">;
}

export const SingleSelectFetchFormikField: React.FC<
  ISingleSelectFetchFormikFieldProps
> = ({ fieldConfig, selectConfig }) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelectFetch
      value={field.value}
      onChange={(selection) => helpers.setValue(selection)}
      onClear={() => helpers.setValue(undefined)}
      {...selectConfig}
    />
  );
};
