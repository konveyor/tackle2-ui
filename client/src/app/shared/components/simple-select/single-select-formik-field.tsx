import React from "react";
import { FieldHookConfig, useField } from "formik";

import { ISimpleSelectProps, SimpleSelect } from "./simple-select";

export interface ISingleSelectFormikFieldProps {
  fieldConfig: FieldHookConfig<any>;
  selectConfig: Omit<ISimpleSelectProps, "value" | "onChange" | "onClear">;
}

export const SingleSelectFormikField: React.FC<
  ISingleSelectFormikFieldProps
> = ({ fieldConfig, selectConfig }) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelect
      value={field.value}
      onChange={(selection) => helpers.setValue(selection)}
      onClear={() => helpers.setValue(undefined)}
      {...selectConfig}
    />
  );
};
