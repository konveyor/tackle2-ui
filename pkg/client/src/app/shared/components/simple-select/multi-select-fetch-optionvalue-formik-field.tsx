import React from "react";
import { FieldHookConfig, useField } from "formik";

import {
  ISimpleSelectFetchProps,
  SimpleSelectFetch,
} from "./simple-select-fetch";
import { OptionWithValue } from "./simple-select";

export interface IMultiSelectFetchOptionValueFormikFieldProps<T> {
  fieldConfig: FieldHookConfig<T[]>;
  selectConfig: Omit<
    ISimpleSelectFetchProps,
    "value" | "options" | "onChange" | "onClear"
  >;
  options: T[];
  toOptionWithValue: (option: T) => OptionWithValue<T>;
  isEqual: (a: T, b: T) => boolean;
}

export const MultiSelectFetchOptionValueFormikField = <T extends any>({
  fieldConfig,
  selectConfig,
  options,
  toOptionWithValue,
  isEqual,
}: IMultiSelectFetchOptionValueFormikFieldProps<T>) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelectFetch
      value={field.value.map(toOptionWithValue)}
      options={options.map(toOptionWithValue)}
      onChange={(selection) => {
        const selectionValue = (selection as OptionWithValue<T>).value;

        const currentValue = field.value;

        let nextValue: T[];
        const elementExists = currentValue.find((f: T) => {
          return isEqual(f, selectionValue);
        });

        if (elementExists) {
          nextValue = currentValue.filter(
            (f: T) => !isEqual(f, selectionValue)
          );
        } else {
          nextValue = [...currentValue, selectionValue];
        }

        helpers.setValue(nextValue);
      }}
      onClear={() => helpers.setValue([])}
      {...selectConfig}
    />
  );
};
