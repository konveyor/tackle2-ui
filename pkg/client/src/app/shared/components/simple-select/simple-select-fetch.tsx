import React from "react";
import { AxiosError } from "axios";

import { WarningTriangleIcon } from "@patternfly/react-icons/dist/esm/icons/warning-triangle-icon";

import { getAxiosErrorMessage } from "@app/utils/utils";
import { ISimpleSelectProps, SimpleSelect } from "./simple-select";

export interface ISimpleSelectFetchProps
  extends Omit<ISimpleSelectProps, "toggleIcon" | "customContent"> {
  isFetching: boolean;
  fetchError?: AxiosError;
}

export const SimpleSelectFetch: React.FC<ISimpleSelectFetchProps> = ({
  isFetching,
  fetchError,
  ...props
}) => {
  let customContent;
  if (isFetching) {
    customContent = <div>&nbsp;Loading...</div>;
  } else if (fetchError) {
    customContent = <div>&nbsp;{getAxiosErrorMessage(fetchError)}</div>;
  }

  return (
    <SimpleSelect
      toggleIcon={fetchError && <WarningTriangleIcon />}
      customContent={customContent}
      {...props}
    />
  );
};
