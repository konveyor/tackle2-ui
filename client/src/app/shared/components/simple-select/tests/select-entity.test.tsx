import React from "react";
import { SimpleSelect } from "../simple-select";
import { render } from "@app/test-config/test-utils";

describe("SimpleSelect", () => {
  it("Renders without crashing", () => {
    const wrapper = render(
      <SimpleSelect
        aria-label="my-selection"
        onChange={jest.fn()}
        options={["uno", "dos"]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
