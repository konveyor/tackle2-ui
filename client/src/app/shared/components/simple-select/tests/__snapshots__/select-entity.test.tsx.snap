// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleSelect Renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pf-c-select"
        data-ouia-component-id="OUIA-Generated-Select-single-1"
        data-ouia-component-type="PF4/Select"
        data-ouia-safe="true"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-label="Options menu"
          aria-labelledby=" pf-select-toggle-id-0"
          class="pf-c-select__toggle"
          id="pf-select-toggle-id-0"
          type="button"
        >
          <div
            class="pf-c-select__toggle-wrapper"
          >
            <span
              class="pf-c-select__toggle-text"
            >
              Select...
            </span>
          </div>
          <span
            class="pf-c-select__toggle-arrow"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pf-c-select"
      data-ouia-component-id="OUIA-Generated-Select-single-1"
      data-ouia-component-type="PF4/Select"
      data-ouia-safe="true"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-label="Options menu"
        aria-labelledby=" pf-select-toggle-id-0"
        class="pf-c-select__toggle"
        id="pf-select-toggle-id-0"
        type="button"
      >
        <div
          class="pf-c-select__toggle-wrapper"
        >
          <span
            class="pf-c-select__toggle-text"
          >
            Select...
          </span>
        </div>
        <span
          class="pf-c-select__toggle-arrow"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
